pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentMetadataPushJobRunner.java

```diff
@@ -106,28 +192,123 @@ public void run() {
     } else {
       JavaSparkContext sparkContext = JavaSparkContext.fromSparkContext(SparkContext.getOrCreate());
       JavaRDD<String> pathRDD = sparkContext.parallelize(segmentsToPush, pushParallelism);
-      URI finalOutputDirURI = outputDirURI;
-      // Prevent using lambda expression in Spark to avoid potential serialization exceptions, use inner function
-      // instead.
-      pathRDD.foreach(new VoidFunction<String>() {
-        @Override
-        public void call(String segmentTarPath)
-            throws Exception {
-          PluginManager.get().init();
-          for (PinotFSSpec pinotFSSpec : pinotFSSpecs) {
-            PinotFSFactory
-                .register(pinotFSSpec.getScheme(), pinotFSSpec.getClassName(), new PinotConfiguration(pinotFSSpec));
+
+      if (_spec.getPushJobSpec().isBatchSegmentUpload()) {
```

Comment:
Clarification:
Is there a reason why we have this if/else check? we basically call the same code whether batch/no-batch right? 
_spec.getPushJobSpec().isBatchSegmentUpload

foreachPartition applies this on the partition itself (i.e N segments per partition based on pushParallelism), whereas forEach goes through each element?

Reply:
yes thats right, I use foreachPartition when in batch upload = true and forEach when it's false. 
The code looks the same but the difference is `segmentUriToTarPathMap` has multiple segments when in batch and only 1 when in non-batch